### PR TITLE
Login via email address

### PR DESCRIPTION
**Pull request by patrickxb**
Fixes CORE-2017 during username/passphrase provisioning.

r? @maxtaco

### CODE REVIEW

**Comment by maxtaco** [comment]
Found a bug: if you login to a machine that you've previously provisioned, but this time by email rather than UID, you'll get a device already provisioned error.  I guess this is related to the issue you brought up, but it seems like we can fix it pretty easily.

File: go/engine/login_provision.go [hunk 1]
```diff
@@ -485,14 +496,47 @@ func (e *LoginProvision) makeDeviceKeys(ctx *Context, args *DeviceWrapArgs) erro
 	return nil
 }
 
-// loadUser will prompt for username (if not provided) and load the user.
+// promptUsername will ask the user for the username (if
+// necessary).
+func (e *LoginProvision) promptUsername(ctx *Context) error {
```
Comment (maxtaco) [inline go/engine/login_provision.go]: rename to `promptUsernameOrEmail`

File: go/engine/login_provision.go [hunk 2]
```diff
@@ -485,14 +496,47 @@ func (e *LoginProvision) makeDeviceKeys(ctx *Context, args *DeviceWrapArgs) erro
...
+	if err != nil {
+		return err
+	}
+	e.arg.Username = username
+	return nil
+}
+
+// If e.arg.Username looks like an email address, it will get a
+// login session in order to map the email address to a keybase user.
+func (e *LoginProvision) checkUsername(ctx *Context) error {
```
Comment (maxtaco) [inline go/engine/login_provision.go]: rename to `checkEmailOrUsername`

**Comment by maxtaco** [comment]
Maybe in this case it suffices to just provide a nicer error message.  I.e.:

    Please login again via `keybase login existingUsername`

File: go/engine/login_provision.go [hunk 3]
```diff
@@ -485,14 +496,47 @@ func (e *LoginProvision) makeDeviceKeys(ctx *Context, args *DeviceWrapArgs) erro
...
+// necessary).
+func (e *LoginProvision) promptUsername(ctx *Context) error {
+	if len(e.arg.Username) != 0 {
+		return nil
+	}
+	username, err := ctx.LoginUI.GetEmailOrUsername(context.TODO(), 0)
+	if err != nil {
+		return err
+	}
+	e.arg.Username = username
```
Comment (maxtaco) [inline go/engine/login_provision.go]: rename etc to `EmailOrUsername`

File: go/libkb/errors.go [hunk 4]
```diff
@@ -560,11 +560,11 @@ func (h ProtocolDowngradeError) Error() string {
 //=============================================================================
 
 type BadUsernameError struct {
-	n string
+	N string
```
Comment (maxtaco) [inline go/libkb/errors.go]: also worth changing this too...

File: go/engine/login_provision.go [hunk 5]
```diff
@@ -485,14 +496,47 @@ func (e *LoginProvision) makeDeviceKeys(ctx *Context, args *DeviceWrapArgs) erro
...
+	e.G().Log.Debug("username %q looks like an email address, must get login session to get user", e.arg.Username)
+	// need to login with it in order to get the username
+	var afterLogin = func(lctx libkb.LoginContext) error {
+		e.arg.Username = lctx.LocalSession().GetUsername().String()
+		return nil
+	}
+	return e.G().LoginState().VerifyEmailAddress(e.arg.Username, ctx.SecretUI, afterLogin)
+}
+
+// loadUser will load the user by name specified in e.arg.Username.
```
Comment (maxtaco) [inline go/engine/login_provision.go]: change comment here too

File: go/libkb/login_state.go [hunk 6]
```diff
@@ -163,6 +163,15 @@ func (s *LoginState) ExternalFunc(f loginHandler, name string) error {
 	return s.loginHandle(f, nil, name)
 }
 
+func (s *LoginState) VerifyEmailAddress(email string, secretUI SecretUI, after afterFn) (err error) {
+	defer Trace(s.G().Log, "VerifyEmailAddress", func() error { return err })()
```
Comment (maxtaco) [inline go/libkb/login_state.go]: :+1:

File: go/engine/login_current_device.go [hunk 7]
```diff
@@ -34,7 +34,10 @@ func (e *LoginCurrentDevice) Prereqs() Prereqs {
 
 // RequiredUIs returns the required UIs.
 func (e *LoginCurrentDevice) RequiredUIs() []libkb.UIKind {
-	return []libkb.UIKind{}
+	return []libkb.UIKind{
+		libkb.LoginUIKind,
+		libkb.SecretUIKind,
+	}
```
Comment (maxtaco) [inline go/engine/login_current_device.go]: why were these needed as a result of this change?

**Comment by maxtaco** [comment]
Some minor comments and questions.  Biggest one is to maybe make the error message more helpful and also to see what's wrong on linux

**Comment by patrickxb** [comment]
@maxtaco What case does "Maybe in this case it suffices to just provide a nicer error message...." refer to?  It's not at a line number in the PR.

**Comment by maxtaco** [comment]
The bug mentioned at the top of this PR discussion, the "device already provisioned error"

File: go/engine/login_current_device.go (hunk 7 above)
Comment (patrickxb) [inline go/engine/login_current_device.go]: They were missing and while writing a test I noticed it as the test didn't provide them to the engine the first time causing a crash.

**Comment by patrickxb** [comment]
PTAL

**Comment by patrickxb** [comment]
Just rebased, which should get linux test fix.

File: go/client/cmd_login.go [hunk 8]
```diff
@@ -78,6 +78,12 @@ func (c *CmdLogin) ParseArgv(ctx *cli.Context) error {
 
 	if nargs == 1 {
 		c.username = ctx.Args()[0]
+		if libkb.CheckEmail.F(c.username) {
+			return errors.New("Please login again via `keybase login existingUsername`")
```
Comment (maxtaco) [inline go/client/cmd_login.go]: is there no way to plumb back through what the existing username already was?

**Comment by maxtaco** [comment]
The fix I requested didn't make it into this PR, right?  If so, can you make a ticket for it?

**Comment by maxtaco** [comment]
Aside from that, ready to go on this.

**Comment by patrickxb** [comment]
This prevents a user from specifying an email address on the command line (`keybase login xxx@xx.com` will return an error), which is what I thought we discussed and was UX-approved by @malgorithms?  You want me to figure out an existing username in that case?

**Comment by maxtaco** [comment]
no this is totally different... let's talk in person, we're miscommunicating.

**Comment by maxtaco** [comment]
My bad, let's push this to a different ticket.
